src/hosting/api.ts

```diff
@@ -275,3 +276,65 @@ export async function createRelease(site: string, channel: string, version: stri
   );
   return res.body;
 }
+
+/**
+ * Adds channel domain to Firebase Auth list.
+ * @param project the project ID.
+ * @param url the url of the channel.
+ */
+export async function addAuthDomain(project: string, url: string): Promise<string[]> {
+  const domains = await getAuthDomains(project);
+  const domain = url.replace("https://", "");
+  let authDomains = domains || [];
+  if (authDomains.includes(domain)) {
+    return authDomains;
+  }
+  authDomains.push(domain);
+  return await updateAuthDomains(project, authDomains);
+}
+
+export async function getApprovedDomains(project: string, site: string): Promise<string[]> {
+  const channels = await listChannels(project, site);
+  // Create a map of channel domain names
+  const channelMap = channels
+    .map((channel: Channel) => channel.url.replace("https://", ""))
+    .reduce((acc: { [key: string]: boolean }, current: string) => {
+      acc[current] = true;
+      return acc;
+    }, {});
+
+  // match any string that has ${site}--*
+  const siteMatch = new RegExp(`${site}--`, "i");
+  // match any string that ends in firebaseapp.com
+  const firebaseAppMatch = new RegExp(/firebaseapp.com$/);
+  const domains = await getAuthDomains(project);
+  let authDomains: string[] = [];
+
+  domains.forEach((domain: string) => {
+    // include domains that end in *.firebaseapp.com because urls belonging
+    // to the live channel should always be included
+    const endsWithFirebaseApp = firebaseAppMatch.test(domain);
+    if (endsWithFirebaseApp) {
+      authDomains.push(domain);
+      return;
+    }
+    // exclude site domains that have no channels
+    const domainWithNoChannel = siteMatch.test(domain) && !channelMap[domain];
+    if (domainWithNoChannel) {
+      return;
+    }
+    // add all other domains (ex: "localhost", etc)
+    authDomains.push(domain);
+  });
+  return authDomains;
+}
+
+/**
+ * Syncs the state of authorized domains with existing channels.
```

Comment:
Could this doc string (maybe getApprovedDomains also/instead?) be updated to reflect how this sync works? When I read "sync" I expect the two lists to come together completely. In this case, any domain listed in the identity API that is a Firebase Channel that no longer exists is removed, and no additional Channels are added to the list (this last part I wouldn't have expected).

Maybe "sync" is the wrong verb here. What about "prune"? If I've followed correctly now, this only will ever remove expired channels, right? That might help clarify what's going on

Reply:
Can the docs be updated here to describe what is done in cleanAuthState? "Sync" is the word that really traps me here - can the description be made more explicit?